{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieving Articles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Nº of papers: 2449\n",
      "/tmp/ipykernel_272977/1879717520.py:89: TqdmWarning: Iterable length 2449 > 1000 but `chunksize` is not set. This may seriously degrade multiprocess performance. Set `chunksize=1` or more.\n",
      "  citation_counts = process_map(fetch_citation_count, pmids, max_workers=10)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "100%|██████████| 2449/2449 [02:42<00:00, 15.10it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
      "\n",
      "INFO:root:Nº of papers w/ citations in less than 3 months: 34\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Retrieving Articles\n",
    "\n",
    "Strcuture:\n",
    "    1. Imports, Variables, Functions\n",
    "    2. Fetching Recent Papers\n",
    "    3. Fetching Citation Counts\n",
    "\"\"\"\n",
    "\n",
    "# 1. Imports, Variables, Functions\n",
    "# imports\n",
    "import requests\n",
    "from datetime import datetime, timedelta\n",
    "import xml.etree.ElementTree as ET\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "from tqdm.contrib.concurrent import (\n",
    "    process_map,\n",
    ")  # Import process_map for multiprocessing\n",
    "\n",
    "# variables\n",
    "journals = [\n",
    "    \"Bioinformatics\",\n",
    "    \"BMC Bioinformatics\",\n",
    "    \"Nucleic Acids Research\",\n",
    "    \"Genome Research\",\n",
    "    \"Genome Biology\",\n",
    "    \"PLOS Computational Biology\",\n",
    "    \"Journal of Computational Biology\",\n",
    "    \"Journal of Biomedical Informatics\",\n",
    "    \"Journal of Cheminformatics\",\n",
    "    \"Molecular Informatics\",\n",
    "    \"Journal of Chemical Information and Modeling\",\n",
    "    \"Nature Machine Intelligence\",\n",
    "    \"IEEE/ACM Transactions on Computational Biology and Bioinformatics\",\n",
    "    \"Computational and Structural Biotechnology Journal\",\n",
    "    \"Briefings in Bioinformatics\",\n",
    "    \"Systems Biology and Applications\",\n",
    "    \"BioSystems\",\n",
    "    \"Algorithms for Molecular Biology\",\n",
    "    \"BMC Genomics\",\n",
    "    \"BMC Systems Biology\",\n",
    "    \"BMC Evolutionary Biology\",\n",
    "    \"BioData Mining\",\n",
    "    \"PLOS Genetics\",\n",
    "    \"Journal of Proteome Research\",\n",
    "    \"Proteins: Structure, Function, and Bioinformatics\",\n",
    "    \"Molecular Systems Biology\",\n",
    "    \"Cell Systems\",\n",
    "    \"GigaScience\",\n",
    "    \"Nature Biomedical Engineering\",  # While broader, it includes significant bioinformatics research.\n",
    "    \"Patterns (Cell Press)\",  # Focuses on data science that impacts science and society, including bioinformatics.\n",
    "]\n",
    "\n",
    "\n",
    "journal_queries = ['\"' + journal + '\"[Journal]' for journal in journals]\n",
    "\n",
    "# query = 'chemoinformatics OR \"AI biomedicine\"'\n",
    "query = \" OR \".join(journal_queries)\n",
    "\n",
    "\n",
    "# functions\n",
    "def fetch_citation_count(pmid):\n",
    "    \"\"\"\n",
    "    Fetch citation count for a given PMID.\n",
    "    \"\"\"\n",
    "    elink_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\"\n",
    "    params = {\n",
    "        \"dbfrom\": \"pubmed\",\n",
    "        \"db\": \"pubmed\",\n",
    "        \"id\": pmid,\n",
    "        \"linkname\": \"pubmed_pubmed_citedin\",  # Link to citing articles\n",
    "        \"retmode\": \"xml\",\n",
    "    }\n",
    "\n",
    "    response = requests.get(elink_url, params=params)\n",
    "    if response.status_code == 200:\n",
    "        root = ET.fromstring(response.content)\n",
    "        citation_count = len(root.findall(\".//LinkSetDb/Link\"))\n",
    "        return citation_count\n",
    "    return 0  # Return 0 if unable to fetch citation count\n",
    "\n",
    "\n",
    "def fetch_citations_for_papers(papers):\n",
    "    \"\"\"\n",
    "    Fetch citation counts for a list of papers in parallel.\n",
    "    \"\"\"\n",
    "    pmids = [paper[\"pmid\"] for paper in papers]\n",
    "    citation_counts = process_map(fetch_citation_count, pmids, max_workers=10)\n",
    "    for paper, citation_count in zip(papers, citation_counts):\n",
    "        paper[\"citations\"] = citation_count\n",
    "    return papers\n",
    "\n",
    "\n",
    "def fetch_recent_papers(query, days_ago=90, n_articles=10000):\n",
    "    \"\"\"\n",
    "    Fetch recent papers from PubMed based on a query and include citation counts.\n",
    "    \"\"\"\n",
    "    # Calculate date range\n",
    "    end_date = datetime.today()\n",
    "    start_date = end_date - timedelta(days=days_ago)\n",
    "    start_date_str = start_date.strftime(\"%Y/%m/%d\")\n",
    "    end_date_str = end_date.strftime(\"%Y/%m/%d\")\n",
    "\n",
    "    # PubMed API endpoint\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\"\n",
    "    params = {\n",
    "        \"db\": \"pubmed\",\n",
    "        \"term\": f'({query}) AND (\"{start_date_str}\"[Date - Publication] : \"{end_date_str}\"[Date - Publication])',\n",
    "        \"retmax\": n_articles,\n",
    "        \"usehistory\": \"y\",\n",
    "    }\n",
    "\n",
    "    response = requests.get(base_url, params=params)\n",
    "    if response.status_code == 200:\n",
    "        root = ET.fromstring(response.content)\n",
    "        webenv = root.find(\".//WebEnv\").text\n",
    "        query_key = root.find(\".//QueryKey\").text\n",
    "\n",
    "        # Fetch actual paper details using efetch\n",
    "        fetch_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "        fetch_params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"query_key\": query_key,\n",
    "            \"WebEnv\": webenv,\n",
    "            \"retmode\": \"xml\",\n",
    "            \"rettype\": \"abstract\",\n",
    "            \"retmax\": n_articles,\n",
    "        }\n",
    "        fetch_response = requests.get(fetch_url, params=fetch_params)\n",
    "        if fetch_response.status_code == 200:\n",
    "            fetch_root = ET.fromstring(fetch_response.content)\n",
    "            papers = []\n",
    "            for article in fetch_root.findall(\".//PubmedArticle\"):\n",
    "                title = article.find(\".//ArticleTitle\").text\n",
    "                pmid = article.find(\".//PMID\").text\n",
    "                link = f\"https://pubmed.ncbi.nlm.nih.gov/{pmid}/\"\n",
    "\n",
    "                # Extract keywords\n",
    "                keyword_list = article.find(\".//KeywordList\")\n",
    "\n",
    "                keywords = (\n",
    "                    [keyword.text for keyword in keyword_list.findall(\".//Keyword\")]\n",
    "                    if keyword_list is not None\n",
    "                    else []\n",
    "                )\n",
    "\n",
    "                # New code to fetch abstracts\n",
    "                abstract_text = article.find(\".//Abstract/AbstractText\")\n",
    "                abstract = (\n",
    "                    abstract_text.text\n",
    "                    if abstract_text is not None\n",
    "                    else \"Abstract not available\"\n",
    "                )\n",
    "\n",
    "                papers.append(\n",
    "                    {\n",
    "                        \"title\": title,\n",
    "                        \"pmid\": pmid,\n",
    "                        \"link\": link,\n",
    "                        \"keywords\": keywords,\n",
    "                        \"abstract\": abstract,\n",
    "                    }\n",
    "                )\n",
    "            return papers\n",
    "    return \"Failed to fetch papers or no papers found.\"\n",
    "\n",
    "\n",
    "# 2. Fetching Recent Papers\n",
    "papers = fetch_recent_papers(query)\n",
    "logging.info(f\"Nº of papers: {len(papers)}\")\n",
    "\n",
    "# 3. Fetching Citation Counts\n",
    "papers = fetch_citations_for_papers(papers)\n",
    "logging.info(\n",
    "    f\"Nº of papers w/ citations in less than 3 months: {len([p for p in papers if p['citations'] > 0])}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "papers = [p for p in papers if p[\"citations\"] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 16 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    }
   ],
   "source": [
    "# 4. Store Data\n",
    "\"\"\"Store Data so that there is a memory for which papers have been fetched.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "df_papers = pd.DataFrame(papers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>link</th>\n",
       "      <th>keywords</th>\n",
       "      <th>abstract</th>\n",
       "      <th>citations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Computational reproducibility of Jupyter noteb...</td>\n",
       "      <td>38206590</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38206590/</td>\n",
       "      <td>[GitHub, Jupyter notebooks, PubMed Central, Py...</td>\n",
       "      <td>Jupyter notebooks facilitate the bundling of e...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Counterfactual formulation of patient-specific...</td>\n",
       "      <td>38191012</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38191012/</td>\n",
       "      <td>[Causal discovery, Causal inference, Computati...</td>\n",
       "      <td>Root causes of disease intuitively correspond ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Whole-genome long-read sequencing downsampling...</td>\n",
       "      <td>38190646</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38190646/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Advances in long-read sequencing (LRS) technol...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEMIG: prediction of the cis-regulatory motif ...</td>\n",
       "      <td>38189539</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38189539/</td>\n",
       "      <td>[algorithms, chromatin accessibility, cluster ...</td>\n",
       "      <td>Sequence motif discovery algorithms enhance th...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Structural coverage of the human interactome.</td>\n",
       "      <td>38180828</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38180828/</td>\n",
       "      <td>[AlphaFold2, PDB, homology modeling databases,...</td>\n",
       "      <td>Complex biological processes in cells are embe...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hotspot propensity across mutational processes.</td>\n",
       "      <td>38177930</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38177930/</td>\n",
       "      <td>[Mutation Rate Variability, Mutation Rate Vari...</td>\n",
       "      <td>The sparsity of mutations observed across tumo...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Auditing the inference processes of medical-im...</td>\n",
       "      <td>38155295</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38155295/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The inferences of most machine-learning models...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PyRates-A code-generation tool for modeling dy...</td>\n",
       "      <td>38150479</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38150479/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The mathematical study of real-world dynamical...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>An agent-based modeling approach for lung fibr...</td>\n",
       "      <td>38127835</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38127835/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The severity of the COVID-19 pandemic has crea...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PANoptosis signaling enables broad immune resp...</td>\n",
       "      <td>38125299</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38125299/</td>\n",
       "      <td>[Disulfiram, Immune, Network pharmacology, PAN...</td>\n",
       "      <td>Accumulating evidence suggests that regulated ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>The cysteine-rich domain in CENP-A chaperone S...</td>\n",
       "      <td>38084929</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38084929/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Centromeric chromatin plays a crucial role in ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Aberrant landscapes of maternal meiotic crosso...</td>\n",
       "      <td>38071472</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38071472/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Meiotic recombination is crucial for human gen...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Dividing out quantification uncertainty allows...</td>\n",
       "      <td>38059347</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38059347/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Differential expression analysis of RNA-seq is...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FastCAR: fast correction for ambient RNA to fa...</td>\n",
       "      <td>38030970</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38030970/</td>\n",
       "      <td>[Ambient RNA correction, Decontamination, Sing...</td>\n",
       "      <td>Cell type-specific differential gene expressio...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Database resources of the National Center for ...</td>\n",
       "      <td>37994677</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37994677/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The National Center for Biotechnology Informat...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A cell abundance analysis based on efficient P...</td>\n",
       "      <td>37990148</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37990148/</td>\n",
       "      <td>[Differential abundance testing, Human endomet...</td>\n",
       "      <td>Single-cell RNA sequencing (scRNA-seq) is a po...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>iNClusive: a database collecting useful inform...</td>\n",
       "      <td>37986218</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37986218/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The incorporation of non-canonical amino acids...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Metabolomic, Proteomic, and Single-Cell Proteo...</td>\n",
       "      <td>37983312</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37983312/</td>\n",
       "      <td>[KRAS, MRTX1133, metabolomics, proteomics, sin...</td>\n",
       "      <td>Mutations in KRAS are common drivers of human ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Characterization and decontamination of backgr...</td>\n",
       "      <td>37973397</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37973397/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Assays such as CITE-seq can measure the abunda...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A new paradigm for molecular dynamics database...</td>\n",
       "      <td>37953362</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37953362/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Molecular dynamics (MD) simulations are keepin...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>The DO-KB Knowledgebase: a 20-year journey dev...</td>\n",
       "      <td>37953304</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37953304/</td>\n",
       "      <td>[]</td>\n",
       "      <td>In 2003, the Human Disease Ontology (DO, https...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>The Reactome Pathway Knowledgebase 2024.</td>\n",
       "      <td>37941124</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37941124/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The Reactome Knowledgebase (https://reactome.o...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTexas Aptamer Database: the collection and lo...</td>\n",
       "      <td>37904593</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37904593/</td>\n",
       "      <td>[]</td>\n",
       "      <td>A growing interest in aptamer research, as evi...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ViralZone 2024 provides higher-resolution imag...</td>\n",
       "      <td>37897348</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37897348/</td>\n",
       "      <td>[]</td>\n",
       "      <td>ViralZone (http://viralzone.expasy.org) is a k...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Model-based characterization of the equilibriu...</td>\n",
       "      <td>37889042</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37889042/</td>\n",
       "      <td>[]</td>\n",
       "      <td>In metazoans, both transcription initiation an...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FusionPDB: a knowledgebase of human fusion pro...</td>\n",
       "      <td>37870473</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37870473/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Tumorigenic functions due to the formation of ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CaeNDR, the Caenorhabditis Natural Diversity R...</td>\n",
       "      <td>37855690</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37855690/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Studies of model organisms have provided impor...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>LncRNADisease v3.0: an updated database of lon...</td>\n",
       "      <td>37819033</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37819033/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Systematic integration of lncRNA-disease assoc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SingPro: a knowledge base providing single-cel...</td>\n",
       "      <td>37819028</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37819028/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Single-cell proteomics (SCP) has emerged as a ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SORC: an integrated spatial omics resource in ...</td>\n",
       "      <td>37811897</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37811897/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The interactions between tumor cells and the m...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>cfOmics: a cell-free multi-Omics database for ...</td>\n",
       "      <td>37757861</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37757861/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Liquid biopsy has emerged as a promising non-i...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>circAtlas 3.0: a gateway to 3 million curated ...</td>\n",
       "      <td>37739414</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37739414/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Recent studies have demonstrated the important...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Computational optimization of antibody humanne...</td>\n",
       "      <td>37550425</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37550425/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Conventional methods for humanizing animal-der...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BioLiP2: an updated structure database for bio...</td>\n",
       "      <td>37522378</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37522378/</td>\n",
       "      <td>[]</td>\n",
       "      <td>With the progress of structural biology, the P...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                title      pmid  \\\n",
       "0   Computational reproducibility of Jupyter noteb...  38206590   \n",
       "1   Counterfactual formulation of patient-specific...  38191012   \n",
       "2   Whole-genome long-read sequencing downsampling...  38190646   \n",
       "3   CEMIG: prediction of the cis-regulatory motif ...  38189539   \n",
       "4       Structural coverage of the human interactome.  38180828   \n",
       "5     Hotspot propensity across mutational processes.  38177930   \n",
       "6   Auditing the inference processes of medical-im...  38155295   \n",
       "7   PyRates-A code-generation tool for modeling dy...  38150479   \n",
       "8   An agent-based modeling approach for lung fibr...  38127835   \n",
       "9   PANoptosis signaling enables broad immune resp...  38125299   \n",
       "10  The cysteine-rich domain in CENP-A chaperone S...  38084929   \n",
       "11  Aberrant landscapes of maternal meiotic crosso...  38071472   \n",
       "12  Dividing out quantification uncertainty allows...  38059347   \n",
       "13  FastCAR: fast correction for ambient RNA to fa...  38030970   \n",
       "14  Database resources of the National Center for ...  37994677   \n",
       "15  A cell abundance analysis based on efficient P...  37990148   \n",
       "16  iNClusive: a database collecting useful inform...  37986218   \n",
       "17  Metabolomic, Proteomic, and Single-Cell Proteo...  37983312   \n",
       "18  Characterization and decontamination of backgr...  37973397   \n",
       "19  A new paradigm for molecular dynamics database...  37953362   \n",
       "20  The DO-KB Knowledgebase: a 20-year journey dev...  37953304   \n",
       "21           The Reactome Pathway Knowledgebase 2024.  37941124   \n",
       "22  UTexas Aptamer Database: the collection and lo...  37904593   \n",
       "23  ViralZone 2024 provides higher-resolution imag...  37897348   \n",
       "24  Model-based characterization of the equilibriu...  37889042   \n",
       "25  FusionPDB: a knowledgebase of human fusion pro...  37870473   \n",
       "26  CaeNDR, the Caenorhabditis Natural Diversity R...  37855690   \n",
       "27  LncRNADisease v3.0: an updated database of lon...  37819033   \n",
       "28  SingPro: a knowledge base providing single-cel...  37819028   \n",
       "29  SORC: an integrated spatial omics resource in ...  37811897   \n",
       "30  cfOmics: a cell-free multi-Omics database for ...  37757861   \n",
       "31  circAtlas 3.0: a gateway to 3 million curated ...  37739414   \n",
       "32  Computational optimization of antibody humanne...  37550425   \n",
       "33  BioLiP2: an updated structure database for bio...  37522378   \n",
       "\n",
       "                                         link  \\\n",
       "0   https://pubmed.ncbi.nlm.nih.gov/38206590/   \n",
       "1   https://pubmed.ncbi.nlm.nih.gov/38191012/   \n",
       "2   https://pubmed.ncbi.nlm.nih.gov/38190646/   \n",
       "3   https://pubmed.ncbi.nlm.nih.gov/38189539/   \n",
       "4   https://pubmed.ncbi.nlm.nih.gov/38180828/   \n",
       "5   https://pubmed.ncbi.nlm.nih.gov/38177930/   \n",
       "6   https://pubmed.ncbi.nlm.nih.gov/38155295/   \n",
       "7   https://pubmed.ncbi.nlm.nih.gov/38150479/   \n",
       "8   https://pubmed.ncbi.nlm.nih.gov/38127835/   \n",
       "9   https://pubmed.ncbi.nlm.nih.gov/38125299/   \n",
       "10  https://pubmed.ncbi.nlm.nih.gov/38084929/   \n",
       "11  https://pubmed.ncbi.nlm.nih.gov/38071472/   \n",
       "12  https://pubmed.ncbi.nlm.nih.gov/38059347/   \n",
       "13  https://pubmed.ncbi.nlm.nih.gov/38030970/   \n",
       "14  https://pubmed.ncbi.nlm.nih.gov/37994677/   \n",
       "15  https://pubmed.ncbi.nlm.nih.gov/37990148/   \n",
       "16  https://pubmed.ncbi.nlm.nih.gov/37986218/   \n",
       "17  https://pubmed.ncbi.nlm.nih.gov/37983312/   \n",
       "18  https://pubmed.ncbi.nlm.nih.gov/37973397/   \n",
       "19  https://pubmed.ncbi.nlm.nih.gov/37953362/   \n",
       "20  https://pubmed.ncbi.nlm.nih.gov/37953304/   \n",
       "21  https://pubmed.ncbi.nlm.nih.gov/37941124/   \n",
       "22  https://pubmed.ncbi.nlm.nih.gov/37904593/   \n",
       "23  https://pubmed.ncbi.nlm.nih.gov/37897348/   \n",
       "24  https://pubmed.ncbi.nlm.nih.gov/37889042/   \n",
       "25  https://pubmed.ncbi.nlm.nih.gov/37870473/   \n",
       "26  https://pubmed.ncbi.nlm.nih.gov/37855690/   \n",
       "27  https://pubmed.ncbi.nlm.nih.gov/37819033/   \n",
       "28  https://pubmed.ncbi.nlm.nih.gov/37819028/   \n",
       "29  https://pubmed.ncbi.nlm.nih.gov/37811897/   \n",
       "30  https://pubmed.ncbi.nlm.nih.gov/37757861/   \n",
       "31  https://pubmed.ncbi.nlm.nih.gov/37739414/   \n",
       "32  https://pubmed.ncbi.nlm.nih.gov/37550425/   \n",
       "33  https://pubmed.ncbi.nlm.nih.gov/37522378/   \n",
       "\n",
       "                                             keywords  \\\n",
       "0   [GitHub, Jupyter notebooks, PubMed Central, Py...   \n",
       "1   [Causal discovery, Causal inference, Computati...   \n",
       "2                                                  []   \n",
       "3   [algorithms, chromatin accessibility, cluster ...   \n",
       "4   [AlphaFold2, PDB, homology modeling databases,...   \n",
       "5   [Mutation Rate Variability, Mutation Rate Vari...   \n",
       "6                                                  []   \n",
       "7                                                  []   \n",
       "8                                                  []   \n",
       "9   [Disulfiram, Immune, Network pharmacology, PAN...   \n",
       "10                                                 []   \n",
       "11                                                 []   \n",
       "12                                                 []   \n",
       "13  [Ambient RNA correction, Decontamination, Sing...   \n",
       "14                                                 []   \n",
       "15  [Differential abundance testing, Human endomet...   \n",
       "16                                                 []   \n",
       "17  [KRAS, MRTX1133, metabolomics, proteomics, sin...   \n",
       "18                                                 []   \n",
       "19                                                 []   \n",
       "20                                                 []   \n",
       "21                                                 []   \n",
       "22                                                 []   \n",
       "23                                                 []   \n",
       "24                                                 []   \n",
       "25                                                 []   \n",
       "26                                                 []   \n",
       "27                                                 []   \n",
       "28                                                 []   \n",
       "29                                                 []   \n",
       "30                                                 []   \n",
       "31                                                 []   \n",
       "32                                                 []   \n",
       "33                                                 []   \n",
       "\n",
       "                                             abstract  citations  \n",
       "0   Jupyter notebooks facilitate the bundling of e...          1  \n",
       "1   Root causes of disease intuitively correspond ...          1  \n",
       "2   Advances in long-read sequencing (LRS) technol...          1  \n",
       "3   Sequence motif discovery algorithms enhance th...          1  \n",
       "4   Complex biological processes in cells are embe...          1  \n",
       "5   The sparsity of mutations observed across tumo...          1  \n",
       "6   The inferences of most machine-learning models...          1  \n",
       "7   The mathematical study of real-world dynamical...          1  \n",
       "8   The severity of the COVID-19 pandemic has crea...          1  \n",
       "9   Accumulating evidence suggests that regulated ...          1  \n",
       "10  Centromeric chromatin plays a crucial role in ...          1  \n",
       "11  Meiotic recombination is crucial for human gen...          1  \n",
       "12  Differential expression analysis of RNA-seq is...          1  \n",
       "13  Cell type-specific differential gene expressio...          1  \n",
       "14  The National Center for Biotechnology Informat...          1  \n",
       "15  Single-cell RNA sequencing (scRNA-seq) is a po...          1  \n",
       "16  The incorporation of non-canonical amino acids...          1  \n",
       "17  Mutations in KRAS are common drivers of human ...          2  \n",
       "18  Assays such as CITE-seq can measure the abunda...          1  \n",
       "19  Molecular dynamics (MD) simulations are keepin...          1  \n",
       "20  In 2003, the Human Disease Ontology (DO, https...          1  \n",
       "21  The Reactome Knowledgebase (https://reactome.o...          3  \n",
       "22  A growing interest in aptamer research, as evi...          2  \n",
       "23  ViralZone (http://viralzone.expasy.org) is a k...          1  \n",
       "24  In metazoans, both transcription initiation an...          1  \n",
       "25  Tumorigenic functions due to the formation of ...          1  \n",
       "26  Studies of model organisms have provided impor...          1  \n",
       "27  Systematic integration of lncRNA-disease assoc...          1  \n",
       "28  Single-cell proteomics (SCP) has emerged as a ...          1  \n",
       "29  The interactions between tumor cells and the m...          1  \n",
       "30  Liquid biopsy has emerged as a promising non-i...          1  \n",
       "31  Recent studies have demonstrated the important...          1  \n",
       "32  Conventional methods for humanizing animal-der...          2  \n",
       "33  With the progress of structural biology, the P...          5  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local usage LLMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Local usage LLMs\n",
    "\n",
    "Structure:\n",
    "    1. Imports, Variables, Functions\n",
    "    2. Load the Model\n",
    "    3. Answer Prompt Request\n",
    "\"\"\"\n",
    "\n",
    "# 1. Imports, Variables, Functions\n",
    "# imports\n",
    "from transformers import pipeline\n",
    "import os, sys\n",
    "from tqdm import tqdm\n",
    "\n",
    "# variables\n",
    "prompt = \"Summarize the following abstract from a scientific article: %s\"\n",
    "model_path = os.path.join(\"..\", \"data\", \"llama-2-7b-chat.Q4_K_M.gguf\")\n",
    "\n",
    "\n",
    "# functions\n",
    "\n",
    "\n",
    "# 2. Load the Model\n",
    "summarizer = pipeline(\"summarization\", model=\"../bart-large-cnn\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "17it [00:45,  2.64s/it]Your max_length is set to 60, but your input_length is only 59. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=29)\n",
      "34it [01:32,  2.71s/it]\n"
     ]
    }
   ],
   "source": [
    "# 3. Answer Prompt Request\n",
    "abstract_summaries = list()\n",
    "for index, row in tqdm(df_papers.iterrows()):\n",
    "    if row[\"abstract\"] is not None:\n",
    "        abstract_summary = summarizer(\n",
    "            prompt % row[\"abstract\"], max_length=60, min_length=30, do_sample=False\n",
    "        )[0][\"summary_text\"]\n",
    "    else:\n",
    "        abstract_summary = None\n",
    "    abstract_summaries.append(abstract_summary)\n",
    "\n",
    "df_papers[\"abstract_summary\"] = abstract_summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>link</th>\n",
       "      <th>keywords</th>\n",
       "      <th>abstract</th>\n",
       "      <th>citations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Computational reproducibility of Jupyter noteb...</td>\n",
       "      <td>38206590</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38206590/</td>\n",
       "      <td>[GitHub, Jupyter notebooks, PubMed Central, Py...</td>\n",
       "      <td>Jupyter notebooks facilitate the bundling of e...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Counterfactual formulation of patient-specific...</td>\n",
       "      <td>38191012</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38191012/</td>\n",
       "      <td>[Causal discovery, Causal inference, Computati...</td>\n",
       "      <td>Root causes of disease intuitively correspond ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Whole-genome long-read sequencing downsampling...</td>\n",
       "      <td>38190646</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38190646/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Advances in long-read sequencing (LRS) technol...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEMIG: prediction of the cis-regulatory motif ...</td>\n",
       "      <td>38189539</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38189539/</td>\n",
       "      <td>[algorithms, chromatin accessibility, cluster ...</td>\n",
       "      <td>Sequence motif discovery algorithms enhance th...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Structural coverage of the human interactome.</td>\n",
       "      <td>38180828</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38180828/</td>\n",
       "      <td>[AlphaFold2, PDB, homology modeling databases,...</td>\n",
       "      <td>Complex biological processes in cells are embe...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hotspot propensity across mutational processes.</td>\n",
       "      <td>38177930</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38177930/</td>\n",
       "      <td>[Mutation Rate Variability, Mutation Rate Vari...</td>\n",
       "      <td>The sparsity of mutations observed across tumo...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Auditing the inference processes of medical-im...</td>\n",
       "      <td>38155295</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38155295/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The inferences of most machine-learning models...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PyRates-A code-generation tool for modeling dy...</td>\n",
       "      <td>38150479</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38150479/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The mathematical study of real-world dynamical...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>An agent-based modeling approach for lung fibr...</td>\n",
       "      <td>38127835</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38127835/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The severity of the COVID-19 pandemic has crea...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PANoptosis signaling enables broad immune resp...</td>\n",
       "      <td>38125299</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38125299/</td>\n",
       "      <td>[Disulfiram, Immune, Network pharmacology, PAN...</td>\n",
       "      <td>Accumulating evidence suggests that regulated ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>The cysteine-rich domain in CENP-A chaperone S...</td>\n",
       "      <td>38084929</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38084929/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Centromeric chromatin plays a crucial role in ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Aberrant landscapes of maternal meiotic crosso...</td>\n",
       "      <td>38071472</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38071472/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Meiotic recombination is crucial for human gen...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Dividing out quantification uncertainty allows...</td>\n",
       "      <td>38059347</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38059347/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Differential expression analysis of RNA-seq is...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FastCAR: fast correction for ambient RNA to fa...</td>\n",
       "      <td>38030970</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38030970/</td>\n",
       "      <td>[Ambient RNA correction, Decontamination, Sing...</td>\n",
       "      <td>Cell type-specific differential gene expressio...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Database resources of the National Center for ...</td>\n",
       "      <td>37994677</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37994677/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The National Center for Biotechnology Informat...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A cell abundance analysis based on efficient P...</td>\n",
       "      <td>37990148</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37990148/</td>\n",
       "      <td>[Differential abundance testing, Human endomet...</td>\n",
       "      <td>Single-cell RNA sequencing (scRNA-seq) is a po...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>iNClusive: a database collecting useful inform...</td>\n",
       "      <td>37986218</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37986218/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The incorporation of non-canonical amino acids...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Metabolomic, Proteomic, and Single-Cell Proteo...</td>\n",
       "      <td>37983312</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37983312/</td>\n",
       "      <td>[KRAS, MRTX1133, metabolomics, proteomics, sin...</td>\n",
       "      <td>Mutations in KRAS are common drivers of human ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Characterization and decontamination of backgr...</td>\n",
       "      <td>37973397</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37973397/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Assays such as CITE-seq can measure the abunda...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A new paradigm for molecular dynamics database...</td>\n",
       "      <td>37953362</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37953362/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Molecular dynamics (MD) simulations are keepin...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>The DO-KB Knowledgebase: a 20-year journey dev...</td>\n",
       "      <td>37953304</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37953304/</td>\n",
       "      <td>[]</td>\n",
       "      <td>In 2003, the Human Disease Ontology (DO, https...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>The Reactome Pathway Knowledgebase 2024.</td>\n",
       "      <td>37941124</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37941124/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The Reactome Knowledgebase (https://reactome.o...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTexas Aptamer Database: the collection and lo...</td>\n",
       "      <td>37904593</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37904593/</td>\n",
       "      <td>[]</td>\n",
       "      <td>A growing interest in aptamer research, as evi...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ViralZone 2024 provides higher-resolution imag...</td>\n",
       "      <td>37897348</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37897348/</td>\n",
       "      <td>[]</td>\n",
       "      <td>ViralZone (http://viralzone.expasy.org) is a k...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Model-based characterization of the equilibriu...</td>\n",
       "      <td>37889042</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37889042/</td>\n",
       "      <td>[]</td>\n",
       "      <td>In metazoans, both transcription initiation an...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FusionPDB: a knowledgebase of human fusion pro...</td>\n",
       "      <td>37870473</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37870473/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Tumorigenic functions due to the formation of ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CaeNDR, the Caenorhabditis Natural Diversity R...</td>\n",
       "      <td>37855690</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37855690/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Studies of model organisms have provided impor...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>LncRNADisease v3.0: an updated database of lon...</td>\n",
       "      <td>37819033</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37819033/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Systematic integration of lncRNA-disease assoc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SingPro: a knowledge base providing single-cel...</td>\n",
       "      <td>37819028</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37819028/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Single-cell proteomics (SCP) has emerged as a ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SORC: an integrated spatial omics resource in ...</td>\n",
       "      <td>37811897</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37811897/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The interactions between tumor cells and the m...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>cfOmics: a cell-free multi-Omics database for ...</td>\n",
       "      <td>37757861</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37757861/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Liquid biopsy has emerged as a promising non-i...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>circAtlas 3.0: a gateway to 3 million curated ...</td>\n",
       "      <td>37739414</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37739414/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Recent studies have demonstrated the important...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Computational optimization of antibody humanne...</td>\n",
       "      <td>37550425</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37550425/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Conventional methods for humanizing animal-der...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BioLiP2: an updated structure database for bio...</td>\n",
       "      <td>37522378</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37522378/</td>\n",
       "      <td>[]</td>\n",
       "      <td>With the progress of structural biology, the P...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                title      pmid  \\\n",
       "0   Computational reproducibility of Jupyter noteb...  38206590   \n",
       "1   Counterfactual formulation of patient-specific...  38191012   \n",
       "2   Whole-genome long-read sequencing downsampling...  38190646   \n",
       "3   CEMIG: prediction of the cis-regulatory motif ...  38189539   \n",
       "4       Structural coverage of the human interactome.  38180828   \n",
       "5     Hotspot propensity across mutational processes.  38177930   \n",
       "6   Auditing the inference processes of medical-im...  38155295   \n",
       "7   PyRates-A code-generation tool for modeling dy...  38150479   \n",
       "8   An agent-based modeling approach for lung fibr...  38127835   \n",
       "9   PANoptosis signaling enables broad immune resp...  38125299   \n",
       "10  The cysteine-rich domain in CENP-A chaperone S...  38084929   \n",
       "11  Aberrant landscapes of maternal meiotic crosso...  38071472   \n",
       "12  Dividing out quantification uncertainty allows...  38059347   \n",
       "13  FastCAR: fast correction for ambient RNA to fa...  38030970   \n",
       "14  Database resources of the National Center for ...  37994677   \n",
       "15  A cell abundance analysis based on efficient P...  37990148   \n",
       "16  iNClusive: a database collecting useful inform...  37986218   \n",
       "17  Metabolomic, Proteomic, and Single-Cell Proteo...  37983312   \n",
       "18  Characterization and decontamination of backgr...  37973397   \n",
       "19  A new paradigm for molecular dynamics database...  37953362   \n",
       "20  The DO-KB Knowledgebase: a 20-year journey dev...  37953304   \n",
       "21           The Reactome Pathway Knowledgebase 2024.  37941124   \n",
       "22  UTexas Aptamer Database: the collection and lo...  37904593   \n",
       "23  ViralZone 2024 provides higher-resolution imag...  37897348   \n",
       "24  Model-based characterization of the equilibriu...  37889042   \n",
       "25  FusionPDB: a knowledgebase of human fusion pro...  37870473   \n",
       "26  CaeNDR, the Caenorhabditis Natural Diversity R...  37855690   \n",
       "27  LncRNADisease v3.0: an updated database of lon...  37819033   \n",
       "28  SingPro: a knowledge base providing single-cel...  37819028   \n",
       "29  SORC: an integrated spatial omics resource in ...  37811897   \n",
       "30  cfOmics: a cell-free multi-Omics database for ...  37757861   \n",
       "31  circAtlas 3.0: a gateway to 3 million curated ...  37739414   \n",
       "32  Computational optimization of antibody humanne...  37550425   \n",
       "33  BioLiP2: an updated structure database for bio...  37522378   \n",
       "\n",
       "                                         link  \\\n",
       "0   https://pubmed.ncbi.nlm.nih.gov/38206590/   \n",
       "1   https://pubmed.ncbi.nlm.nih.gov/38191012/   \n",
       "2   https://pubmed.ncbi.nlm.nih.gov/38190646/   \n",
       "3   https://pubmed.ncbi.nlm.nih.gov/38189539/   \n",
       "4   https://pubmed.ncbi.nlm.nih.gov/38180828/   \n",
       "5   https://pubmed.ncbi.nlm.nih.gov/38177930/   \n",
       "6   https://pubmed.ncbi.nlm.nih.gov/38155295/   \n",
       "7   https://pubmed.ncbi.nlm.nih.gov/38150479/   \n",
       "8   https://pubmed.ncbi.nlm.nih.gov/38127835/   \n",
       "9   https://pubmed.ncbi.nlm.nih.gov/38125299/   \n",
       "10  https://pubmed.ncbi.nlm.nih.gov/38084929/   \n",
       "11  https://pubmed.ncbi.nlm.nih.gov/38071472/   \n",
       "12  https://pubmed.ncbi.nlm.nih.gov/38059347/   \n",
       "13  https://pubmed.ncbi.nlm.nih.gov/38030970/   \n",
       "14  https://pubmed.ncbi.nlm.nih.gov/37994677/   \n",
       "15  https://pubmed.ncbi.nlm.nih.gov/37990148/   \n",
       "16  https://pubmed.ncbi.nlm.nih.gov/37986218/   \n",
       "17  https://pubmed.ncbi.nlm.nih.gov/37983312/   \n",
       "18  https://pubmed.ncbi.nlm.nih.gov/37973397/   \n",
       "19  https://pubmed.ncbi.nlm.nih.gov/37953362/   \n",
       "20  https://pubmed.ncbi.nlm.nih.gov/37953304/   \n",
       "21  https://pubmed.ncbi.nlm.nih.gov/37941124/   \n",
       "22  https://pubmed.ncbi.nlm.nih.gov/37904593/   \n",
       "23  https://pubmed.ncbi.nlm.nih.gov/37897348/   \n",
       "24  https://pubmed.ncbi.nlm.nih.gov/37889042/   \n",
       "25  https://pubmed.ncbi.nlm.nih.gov/37870473/   \n",
       "26  https://pubmed.ncbi.nlm.nih.gov/37855690/   \n",
       "27  https://pubmed.ncbi.nlm.nih.gov/37819033/   \n",
       "28  https://pubmed.ncbi.nlm.nih.gov/37819028/   \n",
       "29  https://pubmed.ncbi.nlm.nih.gov/37811897/   \n",
       "30  https://pubmed.ncbi.nlm.nih.gov/37757861/   \n",
       "31  https://pubmed.ncbi.nlm.nih.gov/37739414/   \n",
       "32  https://pubmed.ncbi.nlm.nih.gov/37550425/   \n",
       "33  https://pubmed.ncbi.nlm.nih.gov/37522378/   \n",
       "\n",
       "                                             keywords  \\\n",
       "0   [GitHub, Jupyter notebooks, PubMed Central, Py...   \n",
       "1   [Causal discovery, Causal inference, Computati...   \n",
       "2                                                  []   \n",
       "3   [algorithms, chromatin accessibility, cluster ...   \n",
       "4   [AlphaFold2, PDB, homology modeling databases,...   \n",
       "5   [Mutation Rate Variability, Mutation Rate Vari...   \n",
       "6                                                  []   \n",
       "7                                                  []   \n",
       "8                                                  []   \n",
       "9   [Disulfiram, Immune, Network pharmacology, PAN...   \n",
       "10                                                 []   \n",
       "11                                                 []   \n",
       "12                                                 []   \n",
       "13  [Ambient RNA correction, Decontamination, Sing...   \n",
       "14                                                 []   \n",
       "15  [Differential abundance testing, Human endomet...   \n",
       "16                                                 []   \n",
       "17  [KRAS, MRTX1133, metabolomics, proteomics, sin...   \n",
       "18                                                 []   \n",
       "19                                                 []   \n",
       "20                                                 []   \n",
       "21                                                 []   \n",
       "22                                                 []   \n",
       "23                                                 []   \n",
       "24                                                 []   \n",
       "25                                                 []   \n",
       "26                                                 []   \n",
       "27                                                 []   \n",
       "28                                                 []   \n",
       "29                                                 []   \n",
       "30                                                 []   \n",
       "31                                                 []   \n",
       "32                                                 []   \n",
       "33                                                 []   \n",
       "\n",
       "                                             abstract  citations  \n",
       "0   Jupyter notebooks facilitate the bundling of e...          1  \n",
       "1   Root causes of disease intuitively correspond ...          1  \n",
       "2   Advances in long-read sequencing (LRS) technol...          1  \n",
       "3   Sequence motif discovery algorithms enhance th...          1  \n",
       "4   Complex biological processes in cells are embe...          1  \n",
       "5   The sparsity of mutations observed across tumo...          1  \n",
       "6   The inferences of most machine-learning models...          1  \n",
       "7   The mathematical study of real-world dynamical...          1  \n",
       "8   The severity of the COVID-19 pandemic has crea...          1  \n",
       "9   Accumulating evidence suggests that regulated ...          1  \n",
       "10  Centromeric chromatin plays a crucial role in ...          1  \n",
       "11  Meiotic recombination is crucial for human gen...          1  \n",
       "12  Differential expression analysis of RNA-seq is...          1  \n",
       "13  Cell type-specific differential gene expressio...          1  \n",
       "14  The National Center for Biotechnology Informat...          1  \n",
       "15  Single-cell RNA sequencing (scRNA-seq) is a po...          1  \n",
       "16  The incorporation of non-canonical amino acids...          1  \n",
       "17  Mutations in KRAS are common drivers of human ...          2  \n",
       "18  Assays such as CITE-seq can measure the abunda...          1  \n",
       "19  Molecular dynamics (MD) simulations are keepin...          1  \n",
       "20  In 2003, the Human Disease Ontology (DO, https...          1  \n",
       "21  The Reactome Knowledgebase (https://reactome.o...          3  \n",
       "22  A growing interest in aptamer research, as evi...          2  \n",
       "23  ViralZone (http://viralzone.expasy.org) is a k...          1  \n",
       "24  In metazoans, both transcription initiation an...          1  \n",
       "25  Tumorigenic functions due to the formation of ...          1  \n",
       "26  Studies of model organisms have provided impor...          1  \n",
       "27  Systematic integration of lncRNA-disease assoc...          1  \n",
       "28  Single-cell proteomics (SCP) has emerged as a ...          1  \n",
       "29  The interactions between tumor cells and the m...          1  \n",
       "30  Liquid biopsy has emerged as a promising non-i...          1  \n",
       "31  Recent studies have demonstrated the important...          1  \n",
       "32  Conventional methods for humanizing animal-der...          2  \n",
       "33  With the progress of structural biology, the P...          5  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Jupyter notebooks facilitate the bundling of executable code with its documentation and output in one interactive environment. They represent a popular mechanism to document and share computational workflows, including for research publications. The reproducibility of computational aspects of research is a key component of scientific reproduc'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstract_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Jupyter notebooks facilitate the bundling of executable code with its documentation and output in one interactive environment. They represent a popular mechanism to document and share computational workflows, including for research publications. Reproducibility of computational aspects of research is a key component of scientific reproducibility.'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstract_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "tuple indices must be integers or slices, not str",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m df_papers\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[0;32m----> 2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtitle\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;28mprint\u001b[39m(row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mabstract\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m(abstract_summary)\n",
      "\u001b[0;31mTypeError\u001b[0m: tuple indices must be integers or slices, not str"
     ]
    }
   ],
   "source": [
    "for index, row in df_papers.iterrows():\n",
    "    print(row[\"title\"])\n",
    "    print(row[\"abstract\"])\n",
    "    print(abstract_summary)\n",
    "    print(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>link</th>\n",
       "      <th>keywords</th>\n",
       "      <th>abstract</th>\n",
       "      <th>citations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Computational reproducibility of Jupyter noteb...</td>\n",
       "      <td>38206590</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38206590/</td>\n",
       "      <td>[GitHub, Jupyter notebooks, PubMed Central, Py...</td>\n",
       "      <td>Jupyter notebooks facilitate the bundling of e...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Counterfactual formulation of patient-specific...</td>\n",
       "      <td>38191012</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38191012/</td>\n",
       "      <td>[Causal discovery, Causal inference, Computati...</td>\n",
       "      <td>Root causes of disease intuitively correspond ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Whole-genome long-read sequencing downsampling...</td>\n",
       "      <td>38190646</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38190646/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Advances in long-read sequencing (LRS) technol...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEMIG: prediction of the cis-regulatory motif ...</td>\n",
       "      <td>38189539</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38189539/</td>\n",
       "      <td>[algorithms, chromatin accessibility, cluster ...</td>\n",
       "      <td>Sequence motif discovery algorithms enhance th...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Structural coverage of the human interactome.</td>\n",
       "      <td>38180828</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38180828/</td>\n",
       "      <td>[AlphaFold2, PDB, homology modeling databases,...</td>\n",
       "      <td>Complex biological processes in cells are embe...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hotspot propensity across mutational processes.</td>\n",
       "      <td>38177930</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38177930/</td>\n",
       "      <td>[Mutation Rate Variability, Mutation Rate Vari...</td>\n",
       "      <td>The sparsity of mutations observed across tumo...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Auditing the inference processes of medical-im...</td>\n",
       "      <td>38155295</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38155295/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The inferences of most machine-learning models...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PyRates-A code-generation tool for modeling dy...</td>\n",
       "      <td>38150479</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38150479/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The mathematical study of real-world dynamical...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>An agent-based modeling approach for lung fibr...</td>\n",
       "      <td>38127835</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38127835/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The severity of the COVID-19 pandemic has crea...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PANoptosis signaling enables broad immune resp...</td>\n",
       "      <td>38125299</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38125299/</td>\n",
       "      <td>[Disulfiram, Immune, Network pharmacology, PAN...</td>\n",
       "      <td>Accumulating evidence suggests that regulated ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>The cysteine-rich domain in CENP-A chaperone S...</td>\n",
       "      <td>38084929</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38084929/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Centromeric chromatin plays a crucial role in ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Aberrant landscapes of maternal meiotic crosso...</td>\n",
       "      <td>38071472</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38071472/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Meiotic recombination is crucial for human gen...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Dividing out quantification uncertainty allows...</td>\n",
       "      <td>38059347</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38059347/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Differential expression analysis of RNA-seq is...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FastCAR: fast correction for ambient RNA to fa...</td>\n",
       "      <td>38030970</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/38030970/</td>\n",
       "      <td>[Ambient RNA correction, Decontamination, Sing...</td>\n",
       "      <td>Cell type-specific differential gene expressio...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Database resources of the National Center for ...</td>\n",
       "      <td>37994677</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37994677/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The National Center for Biotechnology Informat...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A cell abundance analysis based on efficient P...</td>\n",
       "      <td>37990148</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37990148/</td>\n",
       "      <td>[Differential abundance testing, Human endomet...</td>\n",
       "      <td>Single-cell RNA sequencing (scRNA-seq) is a po...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>iNClusive: a database collecting useful inform...</td>\n",
       "      <td>37986218</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37986218/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The incorporation of non-canonical amino acids...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Metabolomic, Proteomic, and Single-Cell Proteo...</td>\n",
       "      <td>37983312</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37983312/</td>\n",
       "      <td>[KRAS, MRTX1133, metabolomics, proteomics, sin...</td>\n",
       "      <td>Mutations in KRAS are common drivers of human ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Characterization and decontamination of backgr...</td>\n",
       "      <td>37973397</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37973397/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Assays such as CITE-seq can measure the abunda...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A new paradigm for molecular dynamics database...</td>\n",
       "      <td>37953362</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37953362/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Molecular dynamics (MD) simulations are keepin...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>The DO-KB Knowledgebase: a 20-year journey dev...</td>\n",
       "      <td>37953304</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37953304/</td>\n",
       "      <td>[]</td>\n",
       "      <td>In 2003, the Human Disease Ontology (DO, https...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>The Reactome Pathway Knowledgebase 2024.</td>\n",
       "      <td>37941124</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37941124/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The Reactome Knowledgebase (https://reactome.o...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UTexas Aptamer Database: the collection and lo...</td>\n",
       "      <td>37904593</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37904593/</td>\n",
       "      <td>[]</td>\n",
       "      <td>A growing interest in aptamer research, as evi...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ViralZone 2024 provides higher-resolution imag...</td>\n",
       "      <td>37897348</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37897348/</td>\n",
       "      <td>[]</td>\n",
       "      <td>ViralZone (http://viralzone.expasy.org) is a k...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Model-based characterization of the equilibriu...</td>\n",
       "      <td>37889042</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37889042/</td>\n",
       "      <td>[]</td>\n",
       "      <td>In metazoans, both transcription initiation an...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FusionPDB: a knowledgebase of human fusion pro...</td>\n",
       "      <td>37870473</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37870473/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Tumorigenic functions due to the formation of ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CaeNDR, the Caenorhabditis Natural Diversity R...</td>\n",
       "      <td>37855690</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37855690/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Studies of model organisms have provided impor...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>LncRNADisease v3.0: an updated database of lon...</td>\n",
       "      <td>37819033</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37819033/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Systematic integration of lncRNA-disease assoc...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SingPro: a knowledge base providing single-cel...</td>\n",
       "      <td>37819028</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37819028/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Single-cell proteomics (SCP) has emerged as a ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SORC: an integrated spatial omics resource in ...</td>\n",
       "      <td>37811897</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37811897/</td>\n",
       "      <td>[]</td>\n",
       "      <td>The interactions between tumor cells and the m...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>cfOmics: a cell-free multi-Omics database for ...</td>\n",
       "      <td>37757861</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37757861/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Liquid biopsy has emerged as a promising non-i...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>circAtlas 3.0: a gateway to 3 million curated ...</td>\n",
       "      <td>37739414</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37739414/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Recent studies have demonstrated the important...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Computational optimization of antibody humanne...</td>\n",
       "      <td>37550425</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37550425/</td>\n",
       "      <td>[]</td>\n",
       "      <td>Conventional methods for humanizing animal-der...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BioLiP2: an updated structure database for bio...</td>\n",
       "      <td>37522378</td>\n",
       "      <td>https://pubmed.ncbi.nlm.nih.gov/37522378/</td>\n",
       "      <td>[]</td>\n",
       "      <td>With the progress of structural biology, the P...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                title      pmid  \\\n",
       "0   Computational reproducibility of Jupyter noteb...  38206590   \n",
       "1   Counterfactual formulation of patient-specific...  38191012   \n",
       "2   Whole-genome long-read sequencing downsampling...  38190646   \n",
       "3   CEMIG: prediction of the cis-regulatory motif ...  38189539   \n",
       "4       Structural coverage of the human interactome.  38180828   \n",
       "5     Hotspot propensity across mutational processes.  38177930   \n",
       "6   Auditing the inference processes of medical-im...  38155295   \n",
       "7   PyRates-A code-generation tool for modeling dy...  38150479   \n",
       "8   An agent-based modeling approach for lung fibr...  38127835   \n",
       "9   PANoptosis signaling enables broad immune resp...  38125299   \n",
       "10  The cysteine-rich domain in CENP-A chaperone S...  38084929   \n",
       "11  Aberrant landscapes of maternal meiotic crosso...  38071472   \n",
       "12  Dividing out quantification uncertainty allows...  38059347   \n",
       "13  FastCAR: fast correction for ambient RNA to fa...  38030970   \n",
       "14  Database resources of the National Center for ...  37994677   \n",
       "15  A cell abundance analysis based on efficient P...  37990148   \n",
       "16  iNClusive: a database collecting useful inform...  37986218   \n",
       "17  Metabolomic, Proteomic, and Single-Cell Proteo...  37983312   \n",
       "18  Characterization and decontamination of backgr...  37973397   \n",
       "19  A new paradigm for molecular dynamics database...  37953362   \n",
       "20  The DO-KB Knowledgebase: a 20-year journey dev...  37953304   \n",
       "21           The Reactome Pathway Knowledgebase 2024.  37941124   \n",
       "22  UTexas Aptamer Database: the collection and lo...  37904593   \n",
       "23  ViralZone 2024 provides higher-resolution imag...  37897348   \n",
       "24  Model-based characterization of the equilibriu...  37889042   \n",
       "25  FusionPDB: a knowledgebase of human fusion pro...  37870473   \n",
       "26  CaeNDR, the Caenorhabditis Natural Diversity R...  37855690   \n",
       "27  LncRNADisease v3.0: an updated database of lon...  37819033   \n",
       "28  SingPro: a knowledge base providing single-cel...  37819028   \n",
       "29  SORC: an integrated spatial omics resource in ...  37811897   \n",
       "30  cfOmics: a cell-free multi-Omics database for ...  37757861   \n",
       "31  circAtlas 3.0: a gateway to 3 million curated ...  37739414   \n",
       "32  Computational optimization of antibody humanne...  37550425   \n",
       "33  BioLiP2: an updated structure database for bio...  37522378   \n",
       "\n",
       "                                         link  \\\n",
       "0   https://pubmed.ncbi.nlm.nih.gov/38206590/   \n",
       "1   https://pubmed.ncbi.nlm.nih.gov/38191012/   \n",
       "2   https://pubmed.ncbi.nlm.nih.gov/38190646/   \n",
       "3   https://pubmed.ncbi.nlm.nih.gov/38189539/   \n",
       "4   https://pubmed.ncbi.nlm.nih.gov/38180828/   \n",
       "5   https://pubmed.ncbi.nlm.nih.gov/38177930/   \n",
       "6   https://pubmed.ncbi.nlm.nih.gov/38155295/   \n",
       "7   https://pubmed.ncbi.nlm.nih.gov/38150479/   \n",
       "8   https://pubmed.ncbi.nlm.nih.gov/38127835/   \n",
       "9   https://pubmed.ncbi.nlm.nih.gov/38125299/   \n",
       "10  https://pubmed.ncbi.nlm.nih.gov/38084929/   \n",
       "11  https://pubmed.ncbi.nlm.nih.gov/38071472/   \n",
       "12  https://pubmed.ncbi.nlm.nih.gov/38059347/   \n",
       "13  https://pubmed.ncbi.nlm.nih.gov/38030970/   \n",
       "14  https://pubmed.ncbi.nlm.nih.gov/37994677/   \n",
       "15  https://pubmed.ncbi.nlm.nih.gov/37990148/   \n",
       "16  https://pubmed.ncbi.nlm.nih.gov/37986218/   \n",
       "17  https://pubmed.ncbi.nlm.nih.gov/37983312/   \n",
       "18  https://pubmed.ncbi.nlm.nih.gov/37973397/   \n",
       "19  https://pubmed.ncbi.nlm.nih.gov/37953362/   \n",
       "20  https://pubmed.ncbi.nlm.nih.gov/37953304/   \n",
       "21  https://pubmed.ncbi.nlm.nih.gov/37941124/   \n",
       "22  https://pubmed.ncbi.nlm.nih.gov/37904593/   \n",
       "23  https://pubmed.ncbi.nlm.nih.gov/37897348/   \n",
       "24  https://pubmed.ncbi.nlm.nih.gov/37889042/   \n",
       "25  https://pubmed.ncbi.nlm.nih.gov/37870473/   \n",
       "26  https://pubmed.ncbi.nlm.nih.gov/37855690/   \n",
       "27  https://pubmed.ncbi.nlm.nih.gov/37819033/   \n",
       "28  https://pubmed.ncbi.nlm.nih.gov/37819028/   \n",
       "29  https://pubmed.ncbi.nlm.nih.gov/37811897/   \n",
       "30  https://pubmed.ncbi.nlm.nih.gov/37757861/   \n",
       "31  https://pubmed.ncbi.nlm.nih.gov/37739414/   \n",
       "32  https://pubmed.ncbi.nlm.nih.gov/37550425/   \n",
       "33  https://pubmed.ncbi.nlm.nih.gov/37522378/   \n",
       "\n",
       "                                             keywords  \\\n",
       "0   [GitHub, Jupyter notebooks, PubMed Central, Py...   \n",
       "1   [Causal discovery, Causal inference, Computati...   \n",
       "2                                                  []   \n",
       "3   [algorithms, chromatin accessibility, cluster ...   \n",
       "4   [AlphaFold2, PDB, homology modeling databases,...   \n",
       "5   [Mutation Rate Variability, Mutation Rate Vari...   \n",
       "6                                                  []   \n",
       "7                                                  []   \n",
       "8                                                  []   \n",
       "9   [Disulfiram, Immune, Network pharmacology, PAN...   \n",
       "10                                                 []   \n",
       "11                                                 []   \n",
       "12                                                 []   \n",
       "13  [Ambient RNA correction, Decontamination, Sing...   \n",
       "14                                                 []   \n",
       "15  [Differential abundance testing, Human endomet...   \n",
       "16                                                 []   \n",
       "17  [KRAS, MRTX1133, metabolomics, proteomics, sin...   \n",
       "18                                                 []   \n",
       "19                                                 []   \n",
       "20                                                 []   \n",
       "21                                                 []   \n",
       "22                                                 []   \n",
       "23                                                 []   \n",
       "24                                                 []   \n",
       "25                                                 []   \n",
       "26                                                 []   \n",
       "27                                                 []   \n",
       "28                                                 []   \n",
       "29                                                 []   \n",
       "30                                                 []   \n",
       "31                                                 []   \n",
       "32                                                 []   \n",
       "33                                                 []   \n",
       "\n",
       "                                             abstract  citations  \n",
       "0   Jupyter notebooks facilitate the bundling of e...          1  \n",
       "1   Root causes of disease intuitively correspond ...          1  \n",
       "2   Advances in long-read sequencing (LRS) technol...          1  \n",
       "3   Sequence motif discovery algorithms enhance th...          1  \n",
       "4   Complex biological processes in cells are embe...          1  \n",
       "5   The sparsity of mutations observed across tumo...          1  \n",
       "6   The inferences of most machine-learning models...          1  \n",
       "7   The mathematical study of real-world dynamical...          1  \n",
       "8   The severity of the COVID-19 pandemic has crea...          1  \n",
       "9   Accumulating evidence suggests that regulated ...          1  \n",
       "10  Centromeric chromatin plays a crucial role in ...          1  \n",
       "11  Meiotic recombination is crucial for human gen...          1  \n",
       "12  Differential expression analysis of RNA-seq is...          1  \n",
       "13  Cell type-specific differential gene expressio...          1  \n",
       "14  The National Center for Biotechnology Informat...          1  \n",
       "15  Single-cell RNA sequencing (scRNA-seq) is a po...          1  \n",
       "16  The incorporation of non-canonical amino acids...          1  \n",
       "17  Mutations in KRAS are common drivers of human ...          2  \n",
       "18  Assays such as CITE-seq can measure the abunda...          1  \n",
       "19  Molecular dynamics (MD) simulations are keepin...          1  \n",
       "20  In 2003, the Human Disease Ontology (DO, https...          1  \n",
       "21  The Reactome Knowledgebase (https://reactome.o...          3  \n",
       "22  A growing interest in aptamer research, as evi...          2  \n",
       "23  ViralZone (http://viralzone.expasy.org) is a k...          1  \n",
       "24  In metazoans, both transcription initiation an...          1  \n",
       "25  Tumorigenic functions due to the formation of ...          1  \n",
       "26  Studies of model organisms have provided impor...          1  \n",
       "27  Systematic integration of lncRNA-disease assoc...          1  \n",
       "28  Single-cell proteomics (SCP) has emerged as a ...          1  \n",
       "29  The interactions between tumor cells and the m...          1  \n",
       "30  Liquid biopsy has emerged as a promising non-i...          1  \n",
       "31  Recent studies have demonstrated the important...          1  \n",
       "32  Conventional methods for humanizing animal-der...          2  \n",
       "33  With the progress of structural biology, the P...          5  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Posting to Twitter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Response(data={'edit_history_tweet_ids': ['1759507685110645060'], 'id': '1759507685110645060', 'text': 'Hellos, world!'}, includes={}, errors=[], meta={})"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Posting to Twitter\n",
    "\n",
    "Structure:\n",
    "    1. Imports, Variables, Functions\n",
    "    2. Authenticate to Twitter\n",
    "    3. Posting to Twitter\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# 1. Imports, Variables, Functions\n",
    "# imports\n",
    "import tweepy\n",
    "import logging\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "# variables\n",
    "api_key = os.getenv(\"API_KEY\")\n",
    "api_key_secret = os.getenv(\"API_KEY_SECRET\")\n",
    "bearer_token = os.getenv(\"BEARER_TOKEN\")\n",
    "access_token = os.getenv(\"ACCESS_TOKEN\")\n",
    "access_token_secret = os.getenv(\"ACCESS_TOKEN_SECRET\")\n",
    "\n",
    "# functions\n",
    "\n",
    "\n",
    "# 2. Authenticate to Twitter\n",
    "client = tweepy.Client(\n",
    "    consumer_key=api_key,\n",
    "    consumer_secret=api_key_secret,\n",
    "    access_token=access_token,\n",
    "    access_token_secret=access_token_secret,\n",
    ")\n",
    "auth = tweepy.OAuth1UserHandler(\n",
    "    api_key, api_key_secret, access_token, access_token_secret\n",
    ")\n",
    "api = tweepy.API(auth)\n",
    "\n",
    "# 3. Posting to Twitter\n",
    "\n",
    "client.create_tweet(text=\"Hellos, world!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing Area"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_recent_papers(query, days_ago=1000, n_articles=10):\n",
    "    \"\"\"\n",
    "    Fetch recent papers from PubMed based on a query and include citation counts.\n",
    "    \"\"\"\n",
    "    # Calculate date range\n",
    "    end_date = datetime.today()\n",
    "    start_date = end_date - timedelta(days=days_ago)\n",
    "    start_date_str = start_date.strftime(\"%Y/%m/%d\")\n",
    "    end_date_str = end_date.strftime(\"%Y/%m/%d\")\n",
    "\n",
    "    # PubMed API endpoint\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\"\n",
    "    params = {\n",
    "        \"db\": \"pubmed\",\n",
    "        \"term\": f'({query}) AND (\"{start_date_str}\"[Date - Publication] : \"{end_date_str}\"[Date - Publication])',\n",
    "        \"retmax\": n_articles,\n",
    "        \"usehistory\": \"y\",\n",
    "    }\n",
    "\n",
    "    response = requests.get(base_url, params=params)\n",
    "    if response.status_code == 200:\n",
    "        root = ET.fromstring(response.content)\n",
    "        webenv = root.find(\".//WebEnv\").text\n",
    "        query_key = root.find(\".//QueryKey\").text\n",
    "\n",
    "        # Fetch actual paper details using efetch\n",
    "        fetch_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "        fetch_params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"query_key\": query_key,\n",
    "            \"WebEnv\": webenv,\n",
    "            \"retmode\": \"xml\",\n",
    "            \"rettype\": \"abstract\",\n",
    "            \"retmax\": n_articles,\n",
    "        }\n",
    "        fetch_response = requests.get(fetch_url, params=fetch_params)\n",
    "        if fetch_response.status_code == 200:\n",
    "            fetch_root = ET.fromstring(fetch_response.content)\n",
    "            papers = []\n",
    "            for article in fetch_root.findall(\".//PubmedArticle\"):\n",
    "                title = article.find(\".//ArticleTitle\").text\n",
    "                pmid = article.find(\".//PMID\").text\n",
    "                link = f\"https://pubmed.ncbi.nlm.nih.gov/{pmid}/\"\n",
    "\n",
    "                # Attempt to extract the DOI\n",
    "                article_doi = article.find(\".//ArticleId[@IdType='doi']\")\n",
    "                doi = article_doi.text if article_doi is not None else None\n",
    "\n",
    "                # Construct the link to the full article using DOI\n",
    "                doi_link = (\n",
    "                    f\"https://doi.org/{doi}\" if doi else \"Link not available\"\n",
    "                )\n",
    "\n",
    "                # Extract keywords\n",
    "                keyword_list = article.find(\".//KeywordList\")\n",
    "\n",
    "                keywords = (\n",
    "                    [keyword.text for keyword in keyword_list.findall(\".//Keyword\")]\n",
    "                    if keyword_list is not None\n",
    "                    else []\n",
    "                )\n",
    "\n",
    "                # New code to fetch abstracts\n",
    "                abstract_text = article.find(\".//Abstract/AbstractText\")\n",
    "                abstract = (\n",
    "                    abstract_text.text\n",
    "                    if abstract_text is not None\n",
    "                    else \"Abstract not available\"\n",
    "                )\n",
    "\n",
    "                papers.append(\n",
    "                    {\n",
    "                        \"title\": title,\n",
    "                        \"pmid\": pmid,\n",
    "                        \"link\": link,\n",
    "                        \"doi_link\": doi_link\n",
    "                        \"keywords\": keywords,\n",
    "                        \"abstract\": abstract,\n",
    "                    }\n",
    "                )\n",
    "            return papers\n",
    "    return \"Failed to fetch papers or no papers found.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = fetch_recent_papers(query=\"paper\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'title': 'Emergency responses for a health workforce under pressure: Lessons learned from system responses to the first wave of the pandemic in Canada.',\n",
       "  'pmid': '38369691',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369691/',\n",
       "  'keywords': ['COVID-19',\n",
       "   'Canada',\n",
       "   'health workforce',\n",
       "   'pandemic response',\n",
       "   'surge capacity'],\n",
       "  'abstract': \"The global health workforce crisis, simmering for decades, was brought to a rolling boil by the COVID-19 pandemic in 2020. With scarce literature, evidence, or best practices to draw from, countries around the world moved to flex their workforces to meet acute challenges of the pandemic, facing demands related to patient volume, patient acuity, and worker vulnerability and absenteeism. One early hypothesis suggested that the acute, short-term pandemic phase would be followed by several waves of resource demands extending over the longer term. However, as the acute phase of the pandemic abated, temporary workforce policies expired and others were repealed with a view of returning to 'normal'. The workforce needs of subsequent phases of pandemic effects were largely ignored despite our new equilibrium resting nowhere near our pre-COVID baseline. In this paper, we describe Canada's early pandemic workforce response. We report the results of an environmental scan of the early workforce strategies adopted in Canada during the first wave of the COVID-19 pandemic. Within an expanded three-part conceptual framework for supporting a sustainable health workforce, we describe 470 strategies and policies that aimed to increase the numbers and flexibility of health workers in Canada, and to maximise their continued availability to work. These strategies targeted all types of health workers and roles, enabling changes to the places health work is done, the way in which care is delivered, and the mechanisms by which it is regulated. Telehealth strategies and virtual care were the most prevalent, followed by role expansion, licensure flexibility, mental health supports for workers, and return to practice of retirees. We explore the degree to which these short-term, acute response strategies might be adapted or extended to support the evolving workforce's long-term needs.\"},\n",
       " {'title': 'Approaches to Design an Efficient, Predictable Global Post-approval Change Management System that Facilitates Continual Improvement and Drug Product Availability.',\n",
       "  'pmid': '38369639',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369639/',\n",
       "  'keywords': ['Drug shortages',\n",
       "   'One-voice-of-quality for post-approval changes (1VQ for PAC)',\n",
       "   'Post-approval changes',\n",
       "   'Quality management',\n",
       "   'Regulatory complexity'],\n",
       "  'abstract': \"The complexity and inter-connectedness of operating in a global world for drug product supply has become an undeniable reality, further underscored by the COVID-19 pandemic. For Post-Approval Changes (PACs) that are an inevitable part of a product's commercial life, the impact of the growing global regulatory complexity and related drug shortages has brought the Global PAC Management\\xa0System to an inflection point in particular for companies that have their products marketed in many countries.This paper illustrates through data analyzed for the first time from 145,000\\u2009+\\u2009PACs for 156 countries, collected by 18 global pharma companies over a 3-year period (2019-2021), how severe the problem of global regulatory complexity is. Only PACs requiring national regulatory agency (NRA) approval prior to implementation were included in the data set. 1 of the 156 country NRAs approved all submitted PACs within a period of 6\\xa0months. The 6-month timeline was chosen because it is the recommended review timeline for major changes in the WHO guidance for vaccines and biotherapeutic products. 10 out of the 156 (6%) countries had no more than 10% of the PACs reviewed and approved in\\u2009>\\u20096 months. In 33 (22%) countries more than half of the PACs took\\u2009>\\u20096 months for approval. It is rare that the same PAC is approved globally within 6 months as individual NRAs take from a few months to years (in some cases\\u2009>\\u20095 years) for their review.The global PAC management complexity has steadily grown over the past 20\\xa0years. Attempts thus far to solve this problem have not made any meaningful difference. Senior leaders and decision-makers across the interdependent components of the complex Global PAC Management System (industry and regulators) must come together and collaboratively manage the problem holistically with the objective of ensuring global drug product availability instead of continuing with distinct stakeholder or country-focused solutions, which can tend to worsen the problem.In this paper, the Chief Quality Officers (CQOs) from 18 of the largest innovator pharma companies (see Acknowledgements) are speaking with One-Voice-of-Quality for PACs (1VQ for PACs Initiative). They are recommending a set of 8 approaches to activate a holistic transformation of the Global PAC Management System. This article presents their view on the problem of global regulatory complexity for managing PACs, it's impact on continual improvement and the risk to drug product supply, as well as approaches that can help alleviate the problem.\"},\n",
       " {'title': 'Accuracy and precision of dried urine spot method for the detection of Schistosoma mansoni circulating cathodic antigens in resource-limited settings.',\n",
       "  'pmid': '38369522',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369522/',\n",
       "  'keywords': ['Accuracy',\n",
       "   'Dried urine spot',\n",
       "   'Point-of-care circulating cathodic antigen',\n",
       "   'Schistosoma mansoni',\n",
       "   'Sensitivity',\n",
       "   'Specificity'],\n",
       "  'abstract': 'The World Health Organization recommends the use of Schisto point-of-care circulating cathodic antigens (Schisto POC-CCA) for screening of Schistosoma mansoni as it offers better sensitivity than microscopy. However, there are limitation facing the use of this method including timely availability of the test cassettes. The aim of this study was to determine the reliability of dried urine spot (DUS) method for collection of urine and detection of S. mansoni using Schisto POC-CCA cassettes in a resource-limited settings.'},\n",
       " {'title': 'Occupational exposure to endotoxins and small cell lung cancer: a systematic review with meta-analysis.',\n",
       "  'pmid': '38369511',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369511/',\n",
       "  'keywords': ['Occupational exposure',\n",
       "   'case–control studies',\n",
       "   'endotoxins',\n",
       "   'meta-analysis',\n",
       "   'small cell carcinoma'],\n",
       "  'abstract': 'The relationship of occupational exposure to endotoxins with different histologic subtypes of lung cancer has not been established. Our objective was to conduct a systematic review with meta-analysis to assess the effect of exposure to endotoxins on the development of small cell lung cancer (SCLC). A bibliographic search was conducted using MEDLINE, Embase, CENTRAL, and Web of Science databases until December 2022, including all cohort and/or case-control studies that examined occupational exposure to endotoxins and SCLC. Risk of bias was assessed using the U.S. Office of Health Assessment and Translation tool. A random effects model was applied, publication bias were assessed, and a sensitivity analysis was conducted. Four papers were selected for meta-analysis purposes. A total of 144 incident cases of SCLC and 897 population or hospital controls were included. Occupational exposure to endotoxins was considered for textile/leather industry and agricultural sector workers exposed to endotoxins originating from wool, cotton, or leather dust. Except for one study, all investigations were classified as having a low probability of risk of biases. The results of the meta-analysis were not statistically significant (pooled OR: 0.86; 95% CI:0.69-1.08). In addition, neither between-study heterogeneity (I'},\n",
       " {'title': 'Making the most of malaria chemoprevention.',\n",
       "  'pmid': '38369497',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369497/',\n",
       "  'keywords': ['Chemoprevention',\n",
       "   'Malaria',\n",
       "   'Policy',\n",
       "   'Seasonal chemoprevention'],\n",
       "  'abstract': 'Against a backdrop of stalled progress in malaria control, it is surprising that the various forms of malaria chemoprevention are not more widely used. The World Health Organization (WHO) has recommended several malaria chemoprevention strategies, some of them for over a decade, and each with documented efficacy and cost effectiveness. In 2022, the WHO updated and augmented its malaria chemoprevention guidelines to facilitate their wider use. This paper considers new insights into the empirical evidence that supports the broader application of chemoprevention and encourages its application as a default strategy for young children living in moderate to high transmission settings given their high risk of severe disease and death. Chemoprevention is an effective medium-term strategy with potential benefits far outweighing costs. There is a strong argument for urgently increasing malaria chemoprevention in endemic countries.'},\n",
       " {'title': 'Gut microbial metabolites SCFAs and chronic kidney disease.',\n",
       "  'pmid': '38369469',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369469/',\n",
       "  'keywords': ['CKD',\n",
       "   'Gut microbial metabolites',\n",
       "   'Inflammatory responses',\n",
       "   'Oxidative stress',\n",
       "   'SCFAs'],\n",
       "  'abstract': \"The global incidence of Chronic Kidney Disease (CKD) is steadily escalating, with discernible linkage to the intricate terrain of intestinal microecology. The intestinal microbiota orchestrates a dynamic equilibrium in the organism, metabolizing dietary-derived compounds, a process which profoundly impacts human health. Among these compounds, short-chain fatty acids (SCFAs), which result from microbial metabolic processes, play a versatile role in influencing host energy homeostasis, immune function, and intermicrobial signaling, etc. SCFAs emerge as pivotal risk factors influencing CKD's development and prognosis. This paper review elucidates the impact of gut microbial metabolites, specifically SCFAs, on CKD, highlighting their role in modulating host inflammatory responses, oxidative stress, cellular autophagy, the immune milieu, and signaling cascades. An in-depth comprehension of the interplay between SCFAs and kidney disease pathogenesis may pave the way for their utilization as biomarkers for CKD progression and prognosis or as novel adjunctive therapeutic strategies.\"},\n",
       " {'title': 'Development of a novel dynamic nosocomial infection risk management method for COVID-19 in outpatient settings.',\n",
       "  'pmid': '38369460',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369460/',\n",
       "  'keywords': ['ARIMA model',\n",
       "   'COVID-19',\n",
       "   'Infection control',\n",
       "   'Outpatient Clinics'],\n",
       "  'abstract': 'Application of accumulated experience and management measures in the prevention and control of coronavirus disease 2019 (COVID-19) has generally depended on the subjective judgment of epidemic intensity, with the quality of prevention and control management being uneven. The present study was designed to develop a novel risk management system for COVID-19 infection in outpatients, with the ability to provide accurate and hierarchical control based on estimated risk of infection.'},\n",
       " {'title': 'Clinical evaluation of posterior restorations over wet and dry dentin using an etch-and-rinse adhesive: A 36-month randomized clinical trial.',\n",
       "  'pmid': '38369403',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369403/',\n",
       "  'keywords': ['Clinical trial',\n",
       "   'Composite restoration',\n",
       "   'Dental adhesive',\n",
       "   'Dental bonding',\n",
       "   'Dentin'],\n",
       "  'abstract': 'To evaluate the clinical performance of posterior restorations over wet and dry dentin with an etch-and-rinse adhesive after 36 months of clinical service.'},\n",
       " {'title': 'TGF-β1 inhibitor P144 protects against benign restenosis after esophageal stenting through TGF-β1/Smads signaling pathway inhibition.',\n",
       "  'pmid': '38369402',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369402/',\n",
       "  'keywords': ['Benign restenosis',\n",
       "   'Esophageal fibroblast',\n",
       "   'Esophageal stenting',\n",
       "   'P144',\n",
       "   'SMAD2/3',\n",
       "   'TGF-β1',\n",
       "   'α-SMA'],\n",
       "  'abstract': \"Esophageal restenosis is a serious complication after esophageal stent placement, which influences the clinical prognosis of stent implantation and the patient's quality of life. TGF-β1/Smads signaling pathway plays an important role in the development of the eosinophilic esophagitis and scar repair after skin trauma. However, the role of TGF-β1/Smads in the development of esophageal restenosis after esophageal stent placement remains unknown. Our study aimed to investigate whether TGF-β1/Smads plays an important role in the development of esophageal restenosis after esophageal stent, and whether the exogenous TGF-β1 inhibitor supplement could ameliorate the esophageal restenosis after esophageal stent.\"},\n",
       " {'title': 'Lassa fever vaccine use cases and demand: Perspectives from select West African experts.',\n",
       "  'pmid': '38369392',\n",
       "  'link': 'https://pubmed.ncbi.nlm.nih.gov/38369392/',\n",
       "  'keywords': ['Hemorrhagic fevers', 'Vaccine use case', 'Zoonoses'],\n",
       "  'abstract': 'Lassa fever (LF) is a zoonotic viral hemorrhagic disease endemic to several West African countries. Approximately 300-500,000 cases occur annually across all ages with 10-20% case fatality rates. A LF vaccine is a recognized public health priority, with several candidates entering clinical trials. However, the perspectives of regional experts regarding critical vaccine properties, ideal delivery methods, and priority target populations remain unclear. Using a mixed methods approach with a standardized questionnaire, we individually interviewed 8 West African stakeholders, each with extensive knowledge and experience of LF. They strongly favored the use of a mass, proactive campaign strategy to immunize a wide age range of people in high-risk areas, including pregnant women and health care workers. We estimated that these and other plausible delivery scenarios could result in an initial demand of anywhere from 1 to 100 million doses, with most demand coming from Nigeria. These findings may help inform LF vaccine development and deployment efforts.'}]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "good_reader",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
